#7944 add dockerfile and installation steps

ttmchiou opened the pull request:
This PR specifically adds the following

 1. Added dockerfile folder: This holds relevant dockerfile so users can build locally if desired

I have separated previous implementation details into the following separate PRs
1. Revised INSTALLING.md to incorporate using docker for install workflows. #8550 
I had to re-order some of the steps but hopefully instructions should be clear enough.
I believe you can close your eyes and just copy all the code blocks and it should still work
https://github.com/tenstorrent/tt-metal/pull/8566

2. Added scripts/docker: This folder holds relevant shell scripts to build/install docker and necessary dependencies. Also includes a shell script to bring up and automatically bring user into the docker container to begin development
https://github.com/tenstorrent/tt-metal/pull/8566
https://github.com/tenstorrent/tt-metal/pull/8565

TT-billteng commented on scripts/docker/docker_bringup.sh:
```diff
@@ -0,0 +1,6 @@
+mkdir -p /home/${USER}/workspace
+HOST_UID=${UID} HOST_GID=${GID} docker compose up -d
+
+docker exec -it tt-metal bash
```
does the `-v` parameter work here? If so, we should give user ability to mount their own folder(s)

ttmchiou replied:
No, the -v will not work here but users do have the ability to mount their own folders in the docker-compose.yaml file underneath the 'volumes' section

The flow should be as such
user edits docker-compose.yaml - volumes section

user runs ./docker_bringup.sh

ttmchiou replied:
Will add small comment in the docker-compose file to indicate to users to mount files here.

ttmchiou replied:
update removed docker-compose and all references to it

TT-billteng commented on scripts/docker/docker-compose.yaml:
```diff
@@ -0,0 +1,33 @@
+      - /etc/passwd:/etc/passwd:ro
+      - /etc/shadow:/etc/shadow:ro
+      # Allow for persist file storage
+      - /home/${USER}/workspace:/home/ubuntu:rw
```
when user enters docker, their username changes to ubuntu?

ttmchiou replied:
yes, but they should have the same permissions as the user itself

TT-billteng commented on dockerfile/tt-metal-x86.Dockerfile:
should this be renamed to `Ubuntu-20.04-x86.Dockerfile` ?

ttmchiou replied:
yes, good point, will rename

ttmchiou commented:
@adifrancescoTT 
Feel free to go over the README and see if there's anything confusing to you

olofj commented:
I'll be honest, I find this usage of docker-compose to be... odd. It's usually used to run a service permanently.

Sure, that's how we use Docker internally, to set up a container that you ssh into to do work, but we don't *have* to use it that way.

I have kept a similar Dockerfile locally, and I just launch the build and test from within it, without "moving into" the docker environment. So, I have a launch script that runs the build, like the below. Note: This isn't for sharing directly, since it has paths/dependencies on my paths, etc.

``` bash
#!/bin/bash -x

set -e

export CONTAINER=2404:latest

[ -d /work/.pipcache ] || mkdir /work/.pipcache

function run_docker {
	sudo docker run \
		--rm -t -i \
		-v /work:/work \
		-v /home:/home \
		-v /dev/hugepages-1G:/dev/hugepages-1G \
		-w /work/tt-metal \
		-e TT_METAL_HOME=/work/tt-metal \
		-e TT_METAL_ENV=dev \
		-e ARCH_NAME=grayskull \
		-e PYTHONPATH=/work/tt-metal \
		-e XDG_CACHE_HOME=/work/.pipcache \
		-e DEVICE_CXX=/usr/bin/clang++ \
		-e SILENT=0 \
		--net host \
		--device /dev/tenstorrent/0 \
		--privileged \
		${CONTAINER} \
		"$@"
}

run_docker script -c "make build -k -j$(nproc)" build.out.$(date +%Y%m%d%H%M%S)
run_docker script -c "make tests -k -j$(nproc)" tests.out.$(date +%Y%m%d%H%M%S)
#run_docker script -c ". ./build/python_env/bin/activate ; ./tests/scripts/run_tests.sh --tt-arch grayskull --pipeline-type post_commit" pytest.out.$(date +%Y%m%d%H%M%S)

```

olofj commented:
FWIW, the Dockerfile I've been using is similar but maybe a bit less verbose. This is the Ubuntu 24.04 version:

``` dockerfile
FROM ubuntu:24.04

# Local setup to use a proxy server for all package installs for speedup
#COPY 00aptproxy /etc/apt/apt.conf.d
RUN apt-get update
RUN apt-get install -y --no-install-recommends apt-utils dialog
RUN DEBIAN_FRONTEND=noninteractive apt-get install -y libboost-all-dev sudo zip unzip
RUN DEBIAN_FRONTEND=noninteractive apt-get install -y tmux vim git
RUN apt-get install -y pciutils kmod python3-pip
RUN apt-get install -y python3 python3-dev python3-venv
RUN apt-get install -y curl jq
RUN apt-get install -y software-properties-common build-essential python3-venv libgoogle-glog-dev libyaml-cpp-dev libboost-all-dev libsndfile1 libhwloc-dev
RUN apt-get install -y clang git git-lfs cmake pandoc libtbb-dev libcapstone-dev pkg-config
RUN apt-get install -y python-is-python3
RUN apt-get install -y clang

# Ubuntu 24.04 has a new enough rustc/cargo, so just go with that instead of rustup for now
RUN apt-get install -y rustc cargo

# Doxygen and gtest install
RUN curl -L "https://www.doxygen.nl/files/doxygen-1.9.6.linux.bin.tar.gz" -o doxygen-1.9.6.linux.bin.tar.gz
RUN tar -xvf doxygen-1.9.6.linux.bin.tar.gz
RUN cd doxygen-1.9.6 && sudo make install

RUN curl -L "https://github.com/google/googletest/archive/refs/tags/v1.13.0.tar.gz" -o v1.13.0.tar.gz
RUN tar -xvf v1.13.0.tar.gz
RUN cd googletest-1.13.0 && cmake -DCMAKE_INSTALL_PREFIX=/usr -DBUILD_SHARED_LIBS=ON . && make && sudo make install

# 24.04 comes with a pre-populated ubuntu account, which conflicts with my own default UID,
# so delete it and create one with my own username. This is obviously non-portable.
RUN deluser ubuntu
RUN useradd -m -s /bin/bash -u 1000 olof

USER olof
```

adifrancescoTT commented on INSTALLING.md:
```diff
@@ -56,23 +56,62 @@ sudo -E python3 setup_hugepages.py enable && sudo -E python3 setup_hugepages.py
+[Docker](https://docs.docker.com/engine/install/ubuntu/#install-using-the-repository) version 26.1.0
+[Docker Compose](https://docs.docker.com/compose/install/linux/#install-using-the-repository) version v2.26.1
+
+### Step 4. Install Docker, Docker Compose, and Build Docker Images
```
Review this and following steps numbering

adifrancescoTT commented on INSTALLING.md:
```diff
@@ -56,23 +56,62 @@ sudo -E python3 setup_hugepages.py enable && sudo -E python3 setup_hugepages.py
 ---
 
-### Step 4. Build from source and start using TT-NN and TT-Metalium!
+### Step 4. Clone tt-metal repo
```
Are the steps before this one, and particularly step 1, still required for the Docker path? I see Install "tt-smi/tt-flash/tt-topology" in the Dockerfile so want to make sure

ttmchiou replied:
So the installation Hugepages (Step 3) and tt-kmd drivers must be installed on the host side as this can't be done in docker. The firmware can be flashed using docker too.

In essence, half of Step 1 and Step 3 are required, and Step 2 is not required if developing in docker.

I think a split into two files is probably a better choice here

adifrancescoTT replied:
Thanks for clarifying @ttmchiou and agree with that. We could link the Docker installation steps right at the beginning of this file by adding something similar to what I share here in Italic (please feel free to rephrase):

"These instructions will guide you through the installation of Tenstorrent system tools and drivers, followed by the installation of TT-Metalium and TT-NN.

_The following steps will help you configure a native host development environment. If you prefer to work in a Docker container, please follow this guide: <link_to_docker_steps>._"

ttmchiou replied:
@adifrancescoTT Do you think its better to have it split into two files or just two independent sections but still living in INSTALLING.md?
Option 1: INSTALLING.md + DOCKER_INSTALLING.md
or 
Option 2: INSTALLING.md
but it would look something \
# Intro
for native developemnt, go here <link to Native Developement section>
for docker development, go here <link to Docker Development section>

## Native Development
Instructions

## Docker Development
Instructions

antodifrancisco replied:
Not a super strong preference, but I think I'm leaning towards what you're suggesting (option 2) so go with it if you're also aligned. My main ask is that each section is fully "self-contained", meaning a user can follow it from start to end and achieve their goal without referencing the other one. Thanks!

adifrancescoTT replied:
Not a super strong preference, but I think I'm leaning towards what you're suggesting (option 2) so go with it if you're also aligned. My main ask is that each section is fully "self-contained", meaning a user can follow it from start to end and achieve their goal without referencing the other one. Thanks!

ttmchiou replied:
@adifrancescoTT 
Updated, let me know your thoughts

adifrancescoTT reviewed (commented):
I want to make sure that keeping both installation paths in the same file (vs 2 Develop in Docker and Develop on host files) is the easiest option. If we don't need to go thorugh tt-smi/tt-flash/tt-kmd then I'd recommend separating the two.
If we end up keeping them in the same file, there's a minor typo in the numbering of the steps from 4 on.

ttmchiou commented:
@olofj 
thanks for your input! It was intended to add a docker compose file as a way of just getting a permanent environment setup for users/devs more as a convenience/example rather than forcing users down a particular dev path.

I don't have any strong opinions either way on this so I'm happy to remove docker compose and use your method if you think using it this way is too strange.

adifrancescoTT commented on INSTALLING.md:
````diff
@@ -61,7 +67,7 @@ sudo -E python3 setup_hugepages.py enable && sudo -E python3 setup_hugepages.py
+sudo apt install git git-lfs python3-pip
 ```
 
 3. Clone the repo.
````
@ttmchiou I believe these steps were different, they also look out of order now (1., 3., 2.). Clone the repo should have instructions on how to clone. Actually looking at Step 3 of Docker, that's how this should look like (with environment variables too). Is it possible there was some typo reorganizing the code?

ttmchiou replied:
@adifrancescoTT
Whoops, i accidently added python3-pip as a requirement in the Native Host Installation section.
Other than that though, I haven't made any changes to that section so I'm not sure what you mean by instructions on how to clone?

ttmchiou replied:
moved changes to INSTALLING.md to another ticket.
https://github.com/tenstorrent/tt-metal/issues/8550

adifrancescoTT reviewed (changes requested):
Unless I'm misreading, I think the Native Development section needs some changes, but the overall structure works.

ttmchiou committed:
#7944: Updated Codeowners with install and dockerfile files

ttmchiou committed:
#7944: Add dockerfile and scripts to bringup docker instance of tt-metal

ttmchiou commented:
Split this PR into smaller PRs 
https://github.com/tenstorrent/tt-metal/pull/8566
https://github.com/tenstorrent/tt-metal/pull/8565
@TT-billteng @tt-rkim

tt-rkim commented on CODEOWNERS:
```diff
@@ -162,3 +162,6 @@ tests/python_api_testing/conv/ @tt-nshanker
 scripts/install @tapspatel
+scripts/docker @ttmchiou @TT-billteng @tt-rkim
+
+dockerfile @ttmchiou @TT-billteng @tt-rkim
```
POSIX new line
<img width="85" alt="Screenshot 2024-05-17 at 10 36 50 AM" src="https://github.com/tenstorrent/tt-metal/assets/109366641/9d6cbcca-e28c-4d66-94c5-c55c0390c000">

You gotta fix your editor man, lol

tt-rkim commented on scripts/docker/requirements.txt:
```diff
@@ -0,0 +1,27 @@
+python3.8-venv=3.8.10-0ubuntu1~20.04.9
+
+cargo
+ninja-build
```
new line

tt-rkim replied:
Also damn, I'm wondering if we can capture these requirements in one nice file somewhere rather than copy pasta in

- miw playbooks
- dev deps JSONs in `.github/actions`
- README

probably some other places

tt-rkim commented on scripts/docker/requirements_dev.txt:
```diff
@@ -0,0 +1,5 @@
+sudo
```
Damn `sudo` is a requirement?

Also, can we add `vim`?

ttmchiou replied:
added

JushBJJ replied:
Now I know why vim was in my docker container LOL!

tt-rkim commented on scripts/docker/requirements_dev.txt:
```diff
@@ -0,0 +1,5 @@
+sudo
+nano
+acl=2.2.53-6
+jq=1.6-1ubuntu0.20.04.1
```
no need to pin acl and jq

tt-rkim commented on dockerfile/ubuntu-20.04-x86.Dockerfile:
```diff
@@ -0,0 +1,40 @@
+COPY /scripts/docker/requirements_dev.txt /opt/tt_metal_infra/scripts/docker/requirements_dev.txt
+RUN apt-get -y update \
+    && xargs -a /opt/tt_metal_infra/scripts/docker/requirements_dev.txt apt-get install -y --no-install-recommends \
+    && rm -rf /var/lib/apt/lists/*
```
Why delete this cache?  IS it for something later?

ttmchiou replied:
keeps the docker layer smaller.
Realistically this saves us about 50mb

tt-rkim commented on dockerfile/ubuntu-20.04-x86.Dockerfile:
```diff
@@ -0,0 +1,40 @@
+COPY /scripts /opt/tt_metal_infra/scripts
+COPY build_metal.sh /scripts/build_metal.sh
+
+# ENV TT_METAL_INFRA_DIR=/opt/tt_metal_infra
```
Can we delete this extra commented stuff?

ttmchiou replied:
Will be for future work in including Virtual Env in Docker. This will be integrated but just disabled for future reference

ttmchiou committed:
#7944: Maintain posix compliance, add vim, remove pinned dep versions

removing pinned version of jq and acl
posix lines for CODEOWNERS
adding vim to requirements_dev.txt
